### PR TITLE
Liqun/e2e transformer test

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: Describe your changes.
To add e2e tests for transformer models 

**Motivation and Context**
To have better test coverage

### CODE REVIEW

File: orttraining/orttraining/python/ort_trainer.py [hunk 1]
```diff
@@ -713,6 +714,19 @@ def prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, in
...
             input = input + (internal_loss_scale,)
 
+        # loss_scale input name is needed to call train_step.
+        # loss_scale input name is determined at the time training session is created.
+        # training session creation is delay to train_step.
+        # as a result, loss_scale input name is not available when calling train_step. 
+        # to work around this problem, we have to use the following pattern matching code to get the loss_scaler argument.
+        # this problem can be solved if leter it is decided that loss scaler is always internal.
+        extra_kargs_keys = [key for key in kwargs.keys() if key not in [desc.name_ for desc in input_desc_with_]]
+        if len(input_desc_with_) > len(input) and len(extra_kargs_keys) == 1 and internal_loss_scale is None:
```
Comment (BowenBao) [inline orttraining/orttraining/python/ort_trainer.py:724]: I think the issue can be resolved by another way. The loss scale is only needed after the exported onnx model is processed by ORTTrainer C++ backend, and by that time the session object is created. 

So when exporting onnx model for the first time, we can work without loss scale input. After export is done, and session is created, we have loss scale input name, and everything continues as normal.

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -711,6 +712,20 @@ def prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, in
...
 
+        # loss_scale input name is needed to call train_step.
+        # loss_scale input name is determined at the time training session is created.
+        # training session creation is delay to train_step.
+        # as a result, loss_scale input name is not available when calling train_step. 
+        # to work around this problem, we have to use the following pattern matching code to get the loss_scaler argument.
+        # this problem can be solved if leter it is decided that loss scaler is always internal.
+        # note: when self.loss_fn_ is present, following pattern is invalid. 
+        extra_kargs_keys = [key for key in kwargs.keys() if key not in [desc.name_ for desc in input_desc_with_]]
+        if self.loss_fn_ is None and len(input_desc_with_) > len(input) and len(extra_kargs_keys) == 1 and internal_loss_scale is None:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:723]: `len(extra_kargs_keys) == 1` prevents future extensions to this dict. If it has more than one extra arg, extra_kargs_keys[0] may not work because python dict does not guarantee elements ordering

File: orttraining/orttraining/test/python/orttraining_test_transformers.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+            def get_lr_this_step(global_step):
+                return get_lr(args, global_step)
+            loss_scaler = LossScaler('loss_scale_input_name', True, up_scale_window=2000)
+
+            option_gradient_accumulation_steps = [8]
+            option_fp16 = [True]
+            option_allreduce_post_accumulation = True
+            option_use_internal_get_lr_this_step = False
+            option_use_internal_loss_scaler = True
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_transformers.py:165]: Setting this flag to True, ort_trainer.py:723 is not executed and the logic change is not exercised (` if self.loss_fn_ is None and len(input_desc_with_) > len(input) and len(extra_kargs_keys) == 1 and internal_loss_scale is None:`)

Is that change on ort_trainer.py really necessary or dead code?

File: orttraining/orttraining/python/ort_trainer.py (hunk 1 above)
Comment (liqunfu) [reply orttraining/orttraining/python/ort_trainer.py]: But loss scale is needed at train_step where session is not create.

File: orttraining/orttraining/test/python/orttraining_test_layer_norm_transform.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    id = 0
+    removed_nodes = []
+    layer_norm_nodes = []
+    # Replace with layer norm
+    for node in first_ReduceMean:
+        layer_norm_input = []
+        layer_norm_output = []
+        layer_norm_input.append(node.input[0])
+        node_sub = map_input_Sub[node.output[0]]
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_layer_norm_transform.py:70]: What if any of the below links breaks?

File: orttraining/orttraining/test/python/orttraining_test_model_transform.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    while i < len(model.graph.node):
+        if model.graph.node[i] == node:
+            break;
+        i += 1
+    return i if i < len(model.graph.node) else None;
+
+def replace_input_arg(model, arg, new_arg):
+    for node in model.graph.node:
+        i = 0
+        while i < len(node.input):
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_model_transform.py:34]: `for i in range(len(node.input)):` ?

File: orttraining/orttraining/test/python/orttraining_test_model_transform.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            if node.input[i] == arg:
+                node.input[i] = new_arg
+            i += 1
+
+def find_weight_index(model, name):
+    index = 0
+    for w in model.graph.initializer:
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_model_transform.py:41]: `for index, w in enumerate(model.graph.initializer):`

File: orttraining/orttraining/test/python/orttraining_test_model_transform.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    return result[0] if len(result) == 1 else None
+
+def find_input_as_initializer(model, arg):
+    for initializer in model.graph.initializer:
+        if initializer.name == arg:
+            return initializer
+    return None
+
+def get_node_index(model, node):
+    i = 0
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_model_transform.py:24]:
```python
for i, n in enumerate(model.graph.node):
    if n == node:
        return i
return None
```

File: orttraining/orttraining/test/python/orttraining_test_model_transform.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+       node.name = '%s_%d' %(node.op_type, i)
+       i += 1
+
+def find_output_node(model, arg):
+    result = []
+    for node in model.graph.node:
+        for input in node.input:
+            if input == arg:
+                result.append(node)
+    return result[0] if len(result) == 1 else None
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_model_transform.py:15]: This is finding the one single node that is using `arg` as input.
If `arg` is not used as input by any node, or if there are more than 1 node using `arg` as input, this method returns `None`. Is this expected? What is the use case for this method?

Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_test_model_transform.py:15]: details commented in code. In short, a transpose node can be folded only if it receives a weight that does not feed other nodes.

File: orttraining/orttraining/test/python/orttraining_test_utils.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        _opset_version=12)
+
+    # trainig loop
+    eval_batch = None
+    model.train()
+    for step, batch in enumerate(dataloader):
+        if eval_batch is None:
+            eval_batch = batch
+
+        if not use_internal_get_lr_this_step:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_utils.py:76]: Is there any scenario with external get_lr passed by kwargs that cant be done using the internal get_lr? If not, what do you think of using internal one in which the user doesnt need to do append lr to the batch? it is easier for them and will help migration the the new API which will drop external get_lr

Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_test_utils.py:76]: I do not see a case where external get_lr can cover but an internal one. So with the new design it will reduce use case variations - variations include the use of external loss scaler and learning rate function.

File: orttraining/orttraining/test/python/orttraining_test_bert_postprocess.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+from orttraining_test_model_transform import add_name, fix_transpose, add_expand_shape
+from orttraining_test_layer_norm_transform import layer_norm_transform
+
+def postprocess_model(model):
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/orttraining_test_bert_postprocess.py:4]: We will need to revisit these post process steps later.

File: orttraining/orttraining/python/ort_trainer.py [hunk 11]
```diff
@@ -598,9 +600,12 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
         self.enable_grad_norm_clip_ = enable_grad_norm_clip
         self.frozen_weights_ = frozen_weights
         self.opset_version_ = _opset_version
-        self.loss_scale_input_name = ''
         self.state_dict_ = None
 
+        # use this special string to workaround a corner case that external loss_scale is passed into train_step as kargs.
+        # see prepare_input_and_fetches for more details.
+        self.loss_scale_input_name = 'unset_loss_scale_input_name'
```
Comment (BowenBao) [inline orttraining/orttraining/python/ort_trainer.py:607]: maybe change name to 'default'?
```suggestion
        self.loss_scale_input_name = 'default_loss_scale_input_name'
```

File: orttraining/orttraining/python/ort_trainer.py [hunk 12]
```diff
@@ -724,6 +729,15 @@ def prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, in
...
             input = input + (internal_loss_scale,)
+        elif self.use_mixed_precision:
+            # loss_scale input name is needed to call train_step, for example:
+            #   kwargs[model.loss_scale_input_name] = loss_scale
+            #   outputs = model.train_step(*args, **kwargs)
+            # However, when first time train_step is called model.loss_scale_input_name is not set.
+            # To workaround this problem, we use the special name 'unset_loss_scale_input_name' to indicate 
+            # the loss_scale.
+            if 'unset_loss_scale_input_name' in kwargs.keys():
+                input = input + (kwargs['unset_loss_scale_input_name'],)
```
Comment (BowenBao) [inline orttraining/orttraining/python/ort_trainer.py:740]: same here
